article.md

```diff
@@ -0,0 +1,322 @@
+   
+    verify_eth_signature_uint256{keccak_ptr=keccak_ptr}(hash_uint, r, s, v, eth_address);
+
+    finalize_keccak(keccak_ptr_start=keccak_ptr_start, keccak_ptr_end=keccak_ptr);
+
+    return ();
+}
+```
+It's **VERY** important that we call `finalize_keccak` because of the internal batching done inside it. Failing to call it will make the keccak function unsound - the prover will be able to choose any value as the keccak's result.
+
+## Finalizing
+If everything matches, no assertion fails and `save_connected_addresses` function will be called, which will bind our two addresses together.
+```cairo
+@external
+func add_connection{
+    syscall_ptr : felt*,
+    pedersen_ptr : HashBuiltin*,
+    range_check_ptr,
+    bitwise_ptr : BitwiseBuiltin*
+}(eth_address : felt, starknet_address : felt, domain_hash: Uint256, signature_len: felt, signature: felt*){
+    assert_valid_eth_signature(eth_address, starknet_address, domain_hash, signature_len, signature);
+    save_connected_addresses(eth_address, starknet_address);
+    return ();
+}
+```
+The last thing to do is actually create a connection between `eth_address` and `starknet_address`. As this is an example, we'll be using a basic map between a pair of addresses and a boolean value (`read` function of a storage variable in cairo will return 0 if a key doesn't exist). In our simple case `save_connected_addresses` just calls an internal `write` function of the storage variable.<br/>
+
+```cairo
+%lang starknet
+
+from starkware.cairo.common.cairo_builtins import HashBuiltin
+from starkware.cairo.common.math import assert_nn
+
+
+@storage_var
+func storage(eth_address : felt, starknet_address : felt) -> (exists: felt){
+}
+```
+Bear in mind that we won't be able to list all the Starknet addresses connected to the Ethereum address or vice versa. All we can do is check whether some two addresses are connected or not. If you want some more advanced functions, you'd need to change how the addresses are bound - for example, instead of <br/>`func storage(eth_address : felt, starknet_address : felt) -> (exists: felt)`
+as a storage variable, you could use <br/>
+`func storage(eth_address : felt, starknet_address_index : felt) -> (starknet_address: felt)` which, instead of mapping a pair of `eth_address` and `starknet_address` to a 1, marking the existence of the connection, maps the pair of `eth_address` and index to the `starknet_address` allowing you to iterate through them:
+```
+storage(0x1, 0) = 0x321312
+storage(0x1, 1) = 0x721397
+storage(0x1, 2) = 0 (not set, there are 2 bindings for address 0x1)
+````
+Every bit of code that I've put in here is in [this repository](https://github.com/software-mansion-labs/EIP712-cairo-article) alongside some additional functions and a python script that serves as an Ethereum signing service. Mind that the code has not been audited and should not be used on production without a proper audit. It is only meant to be a demonstration of the concept.
```

Comment:
"should not be used on production without a proper audit" - "**should not be used on production**." (with bold)